File: lib/src/main/java/org/ligi/tracedroid/TraceDroid.java
package org.ligi.tracedroid;

import android.content.Context;

import org.ligi.tracedroid.collecting.TraceDroidMetaInfo;
import org.ligi.tracedroid.collecting.UncaughtExceptionSaver;
import org.ligi.tracedroid.logging.Log;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FilenameFilter;
import java.lang.Thread.UncaughtExceptionHandler;

public class TraceDroid {

    /**
     * registers a UncaughtException handler which saves the stacktrace with
     * some meta-information to the filesystem
     *
     * @param context - the context
     */
    public static void init(Context context) {

        // Gather Info from the Context
        TraceDroidMetaInfo.init(context);

        // get the current handler
        UncaughtExceptionHandler currentHandler = Thread
                .getDefaultUncaughtExceptionHandler();
        if (currentHandler != null) {
            Log.d("current handler class=" + currentHandler.getClass().getName());
        }


        // register our saving handler if not already registered
        if (!(currentHandler instanceof UncaughtExceptionSaver)) {
            // Register default exceptions handler
            UncaughtExceptionSaver exceptionSaver = new UncaughtExceptionSaver(currentHandler);
            Thread.setDefaultUncaughtExceptionHandler(exceptionSaver);
        }
    }

    /**
     * get the dir for the stacktraces and ensure it exists
     *
     * @return
     */
    public static File getEnsuredTraceDir() {
        File trace_files_dir = new File(TraceDroidMetaInfo.getFilesPath());

        // Try to create the files folder if it doesn't exist
        if (!trace_files_dir.exists()) {
            trace_files_dir.mkdirs();
        }

        return trace_files_dir;
    }

    /**
     * get all Stacktrace Files
     *
     * @return all stacktrace files
     */
    public static File[] getStackTraceFiles() {

        FilenameFilter filter = new FilenameFilter() {
            public boolean accept(File dir, String name) {
                return name.endsWith(TraceDroidMetaInfo.getFileSuffix());
            }
        };

        return getEnsuredTraceDir().listFiles(filter);
    }

    public static String getStackTraceText(int limit) {
        Log.d("Searching Exceptions in: " + TraceDroidMetaInfo.getFilesPath());

        String stack_trace_text = "";
        for (File act_file : getStackTraceFiles()) {
            // limit the number of stack traces
            if (limit-- > 0) {
                try {
                    stack_trace_text += "file: " + act_file.toString();
                    stack_trace_text += System.getProperty("line.separator");

                    if (limit-- > 0) {
                        // Read contents of stacktrace file
                        BufferedReader input = new BufferedReader(
                                new FileReader(act_file));
                        try {
                            String line = null;
                            while ((line = input.readLine()) != null) {
                                stack_trace_text += line;
                                stack_trace_text += System
                                        .getProperty("line.separator");
                            }
                        } finally {
                            input.close();
                        }
                    } else
                        stack_trace_text += " discarded by limit";

                } catch (Exception e) {
                    Log.w("problem loading stacktrace", e);
                }
            }
        }

        return stack_trace_text;
    }

    public final static void deleteStacktraceFiles() {
        for (File act_file : getStackTraceFiles()) {
            act_file.delete();
        }
    }

}

File: lib/src/main/java/org/ligi/tracedroid/TraceDroid.kt
package org.ligi.tracedroid

import android.content.Context
import org.ligi.tracedroid.collecting.TraceDroidMetaInfo
import org.ligi.tracedroid.collecting.UncaughtExceptionSaver
import org.ligi.tracedroid.collecting.createMetaInfo
import org.ligi.tracedroid.logging.BufferedLogTree
import timber.log.Timber
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.FilenameFilter

object TraceDroid {

    lateinit var traceDroidMetaInfo: TraceDroidMetaInfo

    private val bufferedLogTree by lazy { BufferedLogTree() }
    /**
     * registers a UncaughtException handler which saves the stacktrace with
     * some meta-information to the filesystem
     *
     * @param context - the context
     */
    fun init(context: Context) {

        traceDroidMetaInfo = createMetaInfo(context)

        Timber.plant(bufferedLogTree)
        // get the current handler
        val currentHandler = Thread.getDefaultUncaughtExceptionHandler()
        if (currentHandler != null) {
            Timber.d("current handler class=" + currentHandler.javaClass.name)


            // register our saving handler if not already registered
            if (currentHandler !is UncaughtExceptionSaver) {
                // Register default exceptions handler
                val exceptionSaver = UncaughtExceptionSaver(traceDroidMetaInfo,currentHandler,bufferedLogTree)
                Thread.setDefaultUncaughtExceptionHandler(exceptionSaver)
            }
        }
    }// Try to create the files folder if it doesn't exist

    /**
     * get the dir for the stacktraces and ensure it exists
     *
     * @return
     */
    private val ensuredTraceDir: File
        get() {
            val traceFilesDir = File(traceDroidMetaInfo.filesPath)

            // Try to create the files folder if it doesn't exist
            if (!traceFilesDir.exists()) {
                traceFilesDir.mkdirs()
            }
            return traceFilesDir
        }

    /**
     * get all Stacktrace Files
     *
     * @return all stacktrace files
     */
    val stackTraceFiles: Array<File>
        get() {
            val filter = FilenameFilter { _, name -> name.endsWith(traceDroidMetaInfo.fileSuffix) }
            return ensuredTraceDir.listFiles(filter) ?: emptyArray()
        }

    fun getStackTraceText(limit: Int): String {
        var linesLeft = limit
        Timber.d("Searching Exceptions in: " + traceDroidMetaInfo.filesPath)
        val stackTraceText = StringBuilder()
        for (act_file in stackTraceFiles) {
            // limit the number of stack traces
            if (linesLeft-- > 0) {
                try {
                    stackTraceText.append("file: ").append(act_file.toString())
                    stackTraceText.append(System.getProperty("line.separator"))
                    if (linesLeft-- > 0) {
                        // Read contents of stacktrace file
                        val input = BufferedReader(
                                FileReader(act_file))
                        input.use { usedInput ->
                            usedInput.readLines().forEach { line ->
                                stackTraceText.append(line)
                                stackTraceText.append(System.getProperty("line.separator"))
                            }
                        }
                    } else stackTraceText.append(" discarded by limit")
                } catch (e: Exception) {
                    Timber.w(e, "problem loading stacktrace")
                }
            }
        }
        return stackTraceText.toString()
    }

    fun getLog() = bufferedLogTree.buffer

    fun deleteStacktraceFiles() {
        for (act_file in stackTraceFiles) {
            act_file.delete()
        }
    }
}